crates/bevy_ecs/src/reflect/entity_commands.rs

```diff
@@ -182,113 +204,55 @@ impl<'w, 's, 'a> ReflectCommandExt for EntityCommands<'w, 's, 'a> {
-pub struct InsertReflect {
-    /// The entity on which the component will be inserted.
-    pub entity: Entity,
-    /// The reflect [Component](crate::component::Component) that will be added to the entity.
-    pub component: Box<dyn Reflect>,
-}
-
-impl Command for InsertReflect {
-    fn apply(self, world: &mut World) {
-        let registry = world.get_resource::<AppTypeRegistry>().unwrap().clone();
-        insert_reflect(world, self.entity, &registry.read(), self.component);
-    }
-}
-
 /// A [`Command`] that adds the boxed reflect component to an entity using the data in the provided
-/// [`Resource`] that implements [`AsRef<TypeRegistry>`].
+/// [`Resource`] that implements [`ReadTypeRegistry`].
 ///
 /// See [`ReflectCommandExt::insert_reflect_with_registry`] for details.
-pub struct InsertReflectWithRegistry<T: Resource + AsRef<TypeRegistry>> {
+pub struct InsertReflectWithRegistry<T: ReadTypeRegistry> {
     /// The entity on which the component will be inserted.
     pub entity: Entity,
     pub _t: PhantomData<T>,
     /// The reflect [Component](crate::component::Component) that will be added to the entity.
     pub component: Box<dyn Reflect>,
 }
 
-impl<T: Resource + AsRef<TypeRegistry>> Command for InsertReflectWithRegistry<T> {
+impl<T: ReadTypeRegistry> Command for InsertReflectWithRegistry<T> {
     fn apply(self, world: &mut World) {
         world.resource_scope(|world, registry: Mut<T>| {
-            let registry: &TypeRegistry = registry.as_ref().as_ref();
-            insert_reflect(world, self.entity, registry, self.component);
+            let registry = registry.type_registry();
+            let type_info = self.component.type_name();
+            let show_entity = self.entity;
+            let Some(mut entity) = world.get_entity_mut(self.entity) else {
+                panic!(
+                    "error[B0003]: Could not insert a reflected component (of type {type_info}) \
+                    for entity {show_entity:?} because it doesn't exist in this World."
+                );
+            };
+            let Some(type_registration) = registry.get_with_name(type_info) else {
+                panic!(
+                    "Can't insert reflect component {type_info} on {show_entity:?} because \
+                    the component {type_info} isn't registered. Try adding it to the type \
+                    registry. with `app.register_type::<{type_info}>()`"
+                );
+            };
+            let Some(reflect_component) = type_registration.data::<ReflectComponent>() else {
+                panic!(
+                    "Can't insert reflect component {type_info} on {show_entity:?} because \
+                    the component's ReflectComponent {type_info} isn't registered. \
+                    Make sure to add the `#[reflect(Component)]` attribute to \
+                    {type_info}'s type declaration"
+                );
+            };
```

Comment:
It's not this PR fault, but I just wanted to note that these error messages could be improved (maybe in a later PR). For example the first one is ambiguous on whether the component type or the entity doesn't exist in the world. Only the first one has an error code, the others lack it. The wording is also inconsistent ("Could not" vs "Can't", "reflect component" vs "reflected component", "{type_info}" vs "(of type {type_info})" etc etc)

Reply:
I did some minor rewording. Definitively not perfect, but it fixes the points you mentioned.